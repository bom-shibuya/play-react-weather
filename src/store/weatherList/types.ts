export interface IWeatherItem {
  dt_text: string
  main: {
    grnd_level: number
    humidity: number
    pressure: number
    sea_level: number
    temp: number
    temp_kf: number
    temp_max: number
    temp_min: number
  }
  weather: Array<{
    description: string
    id: number
    main: string
  }>
}

export type WeatherList = IWeatherItem[]
