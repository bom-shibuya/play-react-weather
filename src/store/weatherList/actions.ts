import { WeatherList } from './types'

export const FETCH_REQUEST = 'FETCH_REQUEST'
export const FETCH_SUCCESS = 'FETCH_SUCCESS'
export const FETCH_FAILURE = 'FETCH_FAILURE'

export const fetchRequest = (payload: { city: string }) => ({
  payload,
  type: FETCH_REQUEST as typeof FETCH_REQUEST
})

export interface IFetchSuccessPayload {
  city: string
  data: WeatherList
}

export const fetchSuccess = (payload: IFetchSuccessPayload) => ({
  payload,
  type: FETCH_SUCCESS as typeof FETCH_SUCCESS
})

export const fetchFailure = (payload: { city: string }) => ({
  payload,
  type: FETCH_FAILURE as typeof FETCH_FAILURE
})

export type Actions =
  | ReturnType<typeof fetchRequest>
  | ReturnType<typeof fetchSuccess>
  | ReturnType<typeof fetchFailure>
