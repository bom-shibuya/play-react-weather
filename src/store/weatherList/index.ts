export * from './actions'
export * from './reducers'
export * from './operations'
export * from './types'
