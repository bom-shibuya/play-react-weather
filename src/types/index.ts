import { ActionCreator } from 'redux'
import { ThunkAction, ThunkDispatch } from 'redux-thunk'
import {
  Actions as CurrentCityAction,
  ICurrentCityState
} from '../store/currentCity'
import {
  Actions as WeatherListAction,
  IWetherListState
} from '../store/weatherList'

export type RootActions = CurrentCityAction | WeatherListAction
export interface IRootState {
  currentCity: ICurrentCityState['currentCity']
  weatherList: IWetherListState[]
}

export type AsyncDispatch = ThunkDispatch<IRootState, undefined, RootActions>
export type AsyncAction = ThunkAction<
  Promise<void>,
  IRootState,
  undefined,
  RootActions
>
export type AsyncActionCreator = ActionCreator<AsyncAction>
