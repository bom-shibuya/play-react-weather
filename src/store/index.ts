import { applyMiddleware, combineReducers, compose, createStore } from 'redux'
import thunk from 'redux-thunk'
import { currentCity } from './currentCity'
import { weatherList } from './weatherList'

const rootReducer = combineReducers({ currentCity, weatherList })
const composeEnhancer =
  (window as any).__REDUX__DEVTOOLS_EXTENSION_COMPOSE__ || compose

const store = createStore(rootReducer, composeEnhancer(applyMiddleware(thunk)))

export default store
