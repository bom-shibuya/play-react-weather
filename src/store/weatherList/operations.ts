import axios from 'axios'
import { AsyncActionCreator, WeatherList } from '../../types'
import { fetchFailure, fetchRequest, fetchSuccess } from './actions'

const API_KEY = process.env.REACT_APP_API_KEY
const END_POINT = 'https://api.openweathermap.org/data/2.5/forecast'
const getRequestUrl = (cityName: string) =>
  `${END_POINT}?q=${cityName},JP&cnt=16&APPID=${API_KEY}`

export const fetchWeather: AsyncActionCreator = (payload: {
  city: string
}) => async dispatch => {
  dispatch(fetchRequest(payload))
  return axios.get(getRequestUrl(payload.city)).then(
    res => {
      const data = res.data as { list: WeatherList }
      dispatch(
        fetchSuccess({
          city: payload.city,
          data: data.list
        })
      )
    },
    _ => {
      dispatch(fetchFailure(payload))
    }
  )
}
