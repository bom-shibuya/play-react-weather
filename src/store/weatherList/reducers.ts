import { WeatherList } from '../../types'
import { Actions, FETCH_FAILURE, FETCH_REQUEST, FETCH_SUCCESS } from './actions'

export const initialState = {
  isError: false,
  isFetching: false,
  lastUpdateAt: null as null | number,
  list: [] as WeatherList
}

export type WeathersState = typeof initialState

export const weathers = (
  state = initialState,
  action: Actions
): WeathersState => {
  switch (action.type) {
    case FETCH_REQUEST:
      return { ...state, isFetching: true }
    case FETCH_SUCCESS:
      return {
        isError: false,
        isFetching: false,
        lastUpdateAt: Date.now(),
        list: [...action.payload.data]
      }
    case FETCH_FAILURE:
      return {
        ...state,
        isError: true,
        lastUpdateAt: Date.now()
      }
    default:
      return state
  }
}

export interface IWetherListState {
  [key: string]: WeathersState
}

export const weatherList = (state = {}, action: Actions): IWetherListState => {
  switch (action.type) {
    case FETCH_REQUEST:
    case FETCH_SUCCESS:
    case FETCH_FAILURE:
      return {
        ...state,
        [action.payload.city]: weathers(initialState, action)
      }
    default:
      return state
  }
}
