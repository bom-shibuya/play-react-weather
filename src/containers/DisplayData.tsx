import React from 'react'
import { connect } from 'react-redux'
import { Contents } from '../components/organisms'
import { fetchWeather, WeathersState } from '../store/weatherList'
import { AsyncDispatch, IRootState } from '../types'

interface IProps {
  cityName: string
  weathersState: WeathersState
  dispatch: AsyncDispatch
}

export const initialCityList = ['Tokyo', 'Osaka', 'Niigata']

class DisplayDateContents extends React.Component<IProps> {
  public componentDidMount(): Promise<void[]> {
    const { dispatch } = this.props
    return Promise.all(
      initialCityList.map(city =>
        dispatch(
          fetchWeather({
            city
          })
        )
      )
    )
  }
  public render() {
    const { cityName, weathersState } = this.props
    return <Contents cityName={cityName} weathersState={weathersState} />
  }
}

interface IOwnProps {
  store?: unknown
}

const mapStateToProps = (state: IRootState, Ownprops: IOwnProps) => ({
  cityName: state.currentCity,
  weathersState: state.weatherList[state.currentCity]
})

export const DisplayData = connect(mapStateToProps)(DisplayDateContents)
