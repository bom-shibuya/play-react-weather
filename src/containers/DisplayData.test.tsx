import axios from 'axios'
import { mount, ReactWrapper, shallow, ShallowWrapper } from 'enzyme'
import React from 'react'
import configureStore, { MockStoreEnhanced } from 'redux-mock-store'
import thunk from 'redux-thunk'
import { WeathersState } from '../store/weatherList'
import { IRootState } from '../types'
import { DisplayData, initialCityList } from './DisplayData'

describe('DisplayData container test', () => {
  const weathersState: WeathersState = {
    isError: false,
    isFetching: false,
    lastUpdateAt: 1545727291732,
    list: []
  }
  const initialState: IRootState = {
    currentCity: 'Tokyo',
    weatherList: {
      Tokyo: weathersState
    }
  }

  interface IProps {
    cityName: string
    weathersState: WeathersState
  }

  let store: MockStoreEnhanced
  let wrapper: ShallowWrapper<IProps>

  beforeEach(() => {
    const mockStore = configureStore([thunk])
    store = mockStore(initialState)
    wrapper = shallow(<DisplayData store={store} />)
  })

  test('should recieve cityName', () => {
    expect(wrapper.props().cityName).toBe('Tokyo')
  })

  test('should recieve weathersState', () => {
    expect(wrapper.props().weathersState).toMatchObject(weathersState)
  })

  test('onMount, fetch initial city list', async () => {
    const res = { data: { list: [] } }
    axios.get = jest.fn(() => Promise.resolve(res))
    store.dispatch = jest.fn()
    const mountWrapper = mount(<DisplayData store={store} />)
    await mountWrapper.instance()
    expect(store.getActions()).toMatchSnapshot()
  })
})
