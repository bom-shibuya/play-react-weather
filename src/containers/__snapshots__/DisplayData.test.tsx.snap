// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisplayData container test onMount, fetch initial city list 1`] = `Array []`;
